common/autoware_auto_perception_rviz_plugin/src/object_detection/predicted_objects_display.cpp

```diff
@@ -23,163 +26,235 @@ namespace rviz_plugins
 {
 namespace object_detection
 {
+void PredictedObjectsDisplay::parallelizedCreateMarkerWorkerThread(int rank)
+{
+  while (true) {                           // keep the thread alive
+    sem_wait(&starting_semaphores[rank]);  // wait for the signal to start
+
+    // Determine the number of objects to be processed by this thread
+    int length = tmp_msg->objects.size();
+    int num_object_per_thread = static_cast<int>(ceil(length / static_cast<float>(num_threads)));
+
+    int first_object_index = rank * num_object_per_thread;
+    int last_object_index = first_object_index + num_object_per_thread;
+    if (last_object_index > length) {
+      last_object_index = length;
+    }
+    std::vector<visualization_msgs::msg::Marker::SharedPtr> thread_makers;
+    for (int i = first_object_index; i < last_object_index; i++) {
+      thread_makers = tackle_one_object(tmp_msg, i, thread_makers);
+    }
+    push_tmp_markers(thread_makers);
+
+    sem_post(&ending_semaphores[rank]);  // signal the end of the job
+  }
+}
+
 PredictedObjectsDisplay::PredictedObjectsDisplay() : ObjectPolygonDisplayBase("tracks")
 {
+  num_threads_property = new rviz_common::properties::IntProperty(
+    "Num Threads", 4, "Num of Threads to use", this);  // by default, use 4 threads
+  max_num_threads = 8;            // hard code the number of threads to be created
+  num_threads = max_num_threads;  // this will be soon modified in the workerThread
   std::thread worker(&PredictedObjectsDisplay::workerThread, this);
   worker.detach();
 }
 
 void PredictedObjectsDisplay::workerThread()
 {
+  // ParallelizedThread Initialization
+  std::vector<std::thread> thread_vector;
+  starting_semaphores = reinterpret_cast<sem_t *>(malloc(max_num_threads * sizeof(sem_t)));
```

Comment:
I am not really an expert on C++ parallel computing. Using semaphores may not be a best practice because we can replace that with mutex + conditional variables, but semaphores make code easier to read. 
You have mentioned in other sections that having too many mutexes is also not good. 
Does that mean it would be better to use Asio to re-organize the code? Any ideas?

Reply:
Yes, that's my suggestion, use a mature library for multithreaded software.
I know that this might be time-consuming, and requires some C++ knowledge - that's why I added few other suggestions, to allow You going in small steps. 
There is a great introduction to [boost::asio by Micheal Caisse](https://youtu.be/rwOv_tw2eA4?si=C4zIA5PcOQMEr2Gl)

I must add, that i really appreciate that You're implementing this optimization. Thank You!

Reply:
@ralwing 
Thank you for your reviewing !
I agree using too mach semaphores and mutex causes bug. But in reality, it might be time-consuming to use advanced multithread library. Also we want to quickly fix our current code which fails to manage detached thread.
So my suggestion in small step is
1. Manage all threads using semaphores (This PR)
2. Sophisticate the code using std::future, boost::asio and etc.